torch/optim/swa.py

```diff
@@ -0,0 +1,371 @@
+                tmp = torch.empty_like(p.data)
+                tmp.copy_(p.data)
+                p.data.copy_(buf)
+                buf.copy_(tmp)
+
+    def step(self, closure=None):
+        r"""Performs a single optimization step.
+
+        In automatic mode also updates SWA running averages.
+        """
+        self._reset_lr_to_swa()
+        loss = self._optimizer.step(closure)
+        for group in self.param_groups:
+            group["step_counter"] += 1
+            steps = group["step_counter"]
+            if self._auto_mode:
+                if steps > group['swa_start'] and steps % group['swa_freq'] == 0:
+                    self.update_swa_group(group)
+        return loss
+
+    def state_dict(self):
+        r"""Returns the state of SWA as a :class:`dict`.
+
+        It contains three entries:
+            * opt_state - a dict holding current optimization state of the base
+                optimizer. Its content differs between optimizer classes.
+            * swa_state - a dict containing current state of SWA. For each
+                optimized variable it contains swa_buffer keeping the running
+                average of the variable
+            * param_groups - a dict containing all parameter groups
+        """
+        opt_state_dict = self._optimizer.state_dict()
+        swa_state = {(id(k) if isinstance(k, torch.Tensor) else k): v
+                     for k, v in self.state.items()}
+        opt_state = opt_state_dict["state"]
+        param_groups = opt_state_dict["param_groups"]
+        return {"opt_state": opt_state, "swa_state": swa_state,
+                "param_groups": param_groups}
+
+    def load_state_dict(self, state_dict):
+        r"""Loads the optimizer state.
+
+        Args:
+            state_dict (dict): SWA optimizer state. Should be an object returned
+                from a call to `state_dict`.
+        """
+        swa_state_dict = {"state": state_dict["swa_state"],
+                          "param_groups": state_dict["param_groups"]}
+        opt_state_dict = {"state": state_dict["opt_state"],
+                          "param_groups": state_dict["param_groups"]}
+        super(SWA, self).load_state_dict(swa_state_dict)
+        self._optimizer.load_state_dict(opt_state_dict)
+        self._opt_state = self._optimizer.state
+
+    def add_param_group(self, param_group):
+        r"""Add a param group to the :class:`Optimizer` s `param_groups`.
+
+        This can be useful when fine tuning a pre-trained network as frozen
+        layers can be made trainable and added to the :class:`Optimizer` as
+        training progresses.
+
+        Args:
+            param_group (dict): Specifies what Tensors should be optimized along
+            with group specific optimization options.
+        """
+        param_group['n_avg'] = 0
+        param_group['step_counter'] = 0
+        self._optimizer.add_param_group(param_group)
+
+    @staticmethod
+    def bn_update(loader, model, device=None):
```

Comment:
So the point of the `bn_update` function is to update the running statistics of batchnorm for a given model. Imagine someone gave you the weights of a network, but not its batchnorm statistics. What we want to do here is to just make a forward pass on all of our training batches with the model. In the function we also change momentum variables in batchnorm to make the convergence of the statistics faster, but that's not essential. 

So if the user has a model / dataloader that's incompatible, with `bn_update` he could just do `opt.swap_swa_sgd()`  in the end of training and do an extra epoch but with just the forward passes.

Reply:
I have been thinking about how we could avoid doing this extra part in the end of training, and we could probably do so if  
1. If we knew when the last epoch of training has started. In that case in the optimizer when `step()` is called for the first time we could just do `swap_swa_sgd` and do nothing when `step` is called. This way, in the last epoch of training we would be making forward passes on every batch using the SWA weights, as desired. 
2. If we could modify the forward pass of the model and after SWA started we could try in the forward pass to sometimes use the current SWA weights in the forward pass instead of the original weights.

Reply:
Both of these solutions aren't looking great in my opinion. The current implementation is more explicit. I think what @vadimkantorov suggested makes sense, as then the user could just specify his own function that does a forward pass on all of the data, and pass it to `bn_update`. I will be happy to move `bn_update` to `nn.utils` if you think it's a good idea.